src/cryptsetup/cryptsetup.c

```diff
@@ -46,6 +46,14 @@
 #define CRYPT_SECTOR_SIZE 512
 #define CRYPT_MAX_SECTOR_SIZE 4096
 
+typedef enum PasswordRegisteredType {
+        PASSWORD_PASSPHRASE,
+        PASSWORD_RECOVERY_KEY,
+        PASSWORD_PASSPHRASE_AND_RECOVERY_KEY,
+        _PASSWORD_REGISTERED_TYPE_MAX,
+        _PASSWORD_REGISTERED_TYPE_INVALID = -EINVAL,
+} PasswordRegisteredType;
```

Comment:
This is all passed within one file, so I don't think it's useful to deal bother with table. Why not just move the enum into the function (to make the code clearer), but return a string? Something like this:
```C
static const char* check_registered_passwords(const char *pretty, struct crypt_device *cd) {
   if (!streq_ptr(crypt_get_type(cd), CRYPT_LUKS2)) {
       ...
       return "passphrase";
  }
   ...
  if (…)
      return "passphrase";
  else
     return "passphrase or recovery key";
}
```

Reply:
I'm not sure how to handle with this approach the case where all the slots are referenced by systemd tokens, but no recovery key is present.